Error during preview text extraction

Since the upgrade to 5.719 I have a problem, that new message are not displayed using IMAP to outlook.office365.com (where other clients do show the messages). I looks as if all messages arrived after the upgrade date are not shown. I do not see any error message, the logcat does not pint to any problem either.
I already remove all data from K9 and re-created the accounts with no success. I also de-installed the app and tried to start from scratch without success.
I am using K9 on a Samsung SM-A520F, no root, no custom ROM, Android 8.

Thanks. The problem is unrelated to IMAP. It's a bug in the code that extracts the preview text.

Developer TODO:
- Find and fix the underlying bug
- Catch unexpected errors during preview extraction and carry on with the message sync

```
com.fsck.k9.mail.MessagingException: Error appending message: [message subject]
	at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1210)
	at com.fsck.k9.mailstore.LocalFolder.access$1100(LocalFolder.java:75)
	at com.fsck.k9.mailstore.LocalFolder$13.doDbWork(LocalFolder.java:1049)
	at com.fsck.k9.mailstore.LocalFolder$13.doDbWork(LocalFolder.java:1044)
	at com.fsck.k9.mailstore.LockableDatabase.execute(LockableDatabase.java:273)
	at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:1044)
	at com.fsck.k9.mailstore.LocalFolder.appendMessages(LocalFolder.java:971)
	at com.fsck.k9.mailstore.K9BackendFolder.savePartialMessage(K9BackendFolder.kt:227)
	at com.fsck.k9.backend.imap.ImapSync.downloadPartial(ImapSync.kt:692)
	at com.fsck.k9.backend.imap.ImapSync.downloadLargeMessages(ImapSync.kt:581)
	at com.fsck.k9.backend.imap.ImapSync.downloadMessages(ImapSync.kt:380)
	at com.fsck.k9.backend.imap.ImapSync.synchronizeMailboxSynchronous(ImapSync.kt:189)
	at com.fsck.k9.backend.imap.ImapSync.sync(ImapSync.kt:32)
	at com.fsck.k9.backend.imap.ImapBackend.sync(ImapBackend.kt:50)
	at com.fsck.k9.controller.MessagingController.syncFolder(MessagingController.java:658)
	at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:615)
	at com.fsck.k9.controller.MessagingController.lambda$synchronizeMailbox$3$MessagingController(MessagingController.java:600)
	at com.fsck.k9.controller.-$$Lambda$MessagingController$Xx9NETpzWcoXEDzz-o_rzXywOEM.run(Unknown Source:8)
	at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:217)
	at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:111)
	at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:160)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.IllegalArgumentException: index: -1; length: 0
	at com.fsck.k9.message.html.EmailSection.get(EmailSection.kt:20)
	at com.fsck.k9.message.html.EmailSection.charAt(EmailSection.kt:8)
	at com.fsck.k9.message.extractors.PreviewTextExtractor.getQuoteHeaderIndex(PreviewTextExtractor.kt:105)
	at com.fsck.k9.message.extractors.PreviewTextExtractor.isQuoteHeaderOnly(PreviewTextExtractor.kt:96)
	at com.fsck.k9.message.extractors.PreviewTextExtractor.extractUnquotedText(PreviewTextExtractor.kt:76)
	at com.fsck.k9.message.extractors.PreviewTextExtractor.stripTextForPreview(PreviewTextExtractor.kt:33)
	at com.fsck.k9.message.extractors.PreviewTextExtractor.extractPreview(PreviewTextExtractor.kt:17)
	at com.fsck.k9.message.extractors.MessagePreviewCreator.createPreview(MessagePreviewCreator.java:33)
	at com.fsck.k9.mailstore.LocalFolder.saveMessage(LocalFolder.java:1131)
	... 21 more
```

--- a/app/core/src/main/java/com/fsck/k9/message/extractors/MessagePreviewCreator.java
+++ b/app/core/src/main/java/com/fsck/k9/message/extractors/MessagePreviewCreator.java
@@ -6,6 +6,8 @@ import androidx.annotation.NonNull;
 import com.fsck.k9.mail.Message;
 import com.fsck.k9.mail.Part;
 
+import timber.log.Timber;
+
 
 public class MessagePreviewCreator {
     private final TextPartFinder textPartFinder;
@@ -33,6 +35,10 @@ public class MessagePreviewCreator {
             String previewText = previewTextExtractor.extractPreview(textPart);
             return PreviewResult.text(previewText);
         } catch (PreviewExtractionException e) {
+            Timber.w(e, "Failed to extract preview text");
+            return PreviewResult.error();
+        } catch (Exception e) {
+            Timber.e(e, "Unexpected error while trying to extract preview text");
             return PreviewResult.error();
         }
     }

--- a/app/core/src/main/java/com/fsck/k9/message/extractors/PreviewTextExtractor.kt
+++ b/app/core/src/main/java/com/fsck/k9/message/extractors/PreviewTextExtractor.kt
@@ -65,10 +65,9 @@ internal class PreviewTextExtractor {
     }
 
     private fun extractUnquotedText(text: String): String {
+        if (text.isEmpty()) return ""
         val emailSections = EmailSectionExtractor.extract(text)
-        if (emailSections.isEmpty()) {
-            return ""
-        }
+        if (emailSections.isEmpty()) return ""
 
         val firstEmailSection = emailSections.first()
         val replySections = if (firstEmailSection.quoteDepth == 0) {
@@ -99,6 +98,8 @@ internal class PreviewTextExtractor {
     private val EmailSection.quoteHeaderIndex: Int
         get() {
             var quoteHeaderIndex = lastIndex
+            if (quoteHeaderIndex == -1) return -1
+
             while (quoteHeaderIndex > 0 && this[quoteHeaderIndex] == '\n') {
                 quoteHeaderIndex--
             }

--- a/app/core/src/test/java/com/fsck/k9/message/extractors/MessagePreviewCreatorTest.java
+++ b/app/core/src/test/java/com/fsck/k9/message/extractors/MessagePreviewCreatorTest.java
@@ -83,6 +83,19 @@ public class MessagePreviewCreatorTest {
         assertEquals(PreviewType.ERROR, result.getPreviewType());
     }
 
+    @Test
+    public void createPreview_withPreviewTextExtractorThrowingUnexpectedException() throws Exception {
+        Message message = createDummyMessage();
+        Part textPart = createTextPart("text/plain");
+        when(textPartFinder.findFirstTextPart(message)).thenReturn(textPart);
+        when(previewTextExtractor.extractPreview(textPart)).thenThrow(new IllegalStateException(""));
+
+        PreviewResult result = previewCreator.createPreview(message);
+
+        assertFalse(result.isPreviewTextAvailable());
+        assertEquals(PreviewType.ERROR, result.getPreviewType());
+    }
+
     private Message createDummyMessage() {
         return new MimeMessage();
     }

--- a/app/core/src/test/java/com/fsck/k9/message/extractors/PreviewTextExtractorTest.kt
+++ b/app/core/src/test/java/com/fsck/k9/message/extractors/PreviewTextExtractorTest.kt
@@ -197,4 +197,14 @@ class PreviewTextExtractorTest {
 
         assertThat(preview).isEqualTo("Reply text")
     }
+
+    @Test
+    fun extractPreview_emptyBody() {
+        val text = ""
+        val part = MessageCreationHelper.createTextPart("text/plain", text)
+
+        val preview = previewTextExtractor.extractPreview(part)
+
+        assertThat(preview).isEqualTo("")
+    }
 }

